pkg/controller/egress/controller.go

```diff
@@ -128,13 +283,111 @@ func (c *EgressController) processNextEgressGroupWorkItem() bool {
 	return true
 }
 
-func (c *EgressController) syncEgressGroup(key string) error {
+func (c *EgressController) getIPAllocation(egressName string) (net.IP, string, bool) {
+	c.ipAllocationMutex.RLock()
+	defer c.ipAllocationMutex.RUnlock()
+	allocation, exists := c.ipAllocationMap[egressName]
+	if !exists {
+		return nil, "", false
+	}
+	return allocation.ip, allocation.ipPool, true
+}
+
+func (c *EgressController) deleteIPAllocation(egressName string) {
+	c.ipAllocationMutex.Lock()
+	defer c.ipAllocationMutex.Unlock()
+	delete(c.ipAllocationMap, egressName)
+}
+
+func (c *EgressController) setIPAllocation(egressName string, ip net.IP, poolName string) {
+	c.ipAllocationMutex.Lock()
+	defer c.ipAllocationMutex.Unlock()
+	c.ipAllocationMap[egressName] = &ipAllocation{
+		ip:     ip,
+		ipPool: poolName,
+	}
+}
+
+// syncEgressIP is responsible for releasing stale EgressIP and allocating new EgressIP for an Egress if applicable.
+func (c *EgressController) syncEgressIP(egress *egressv1alpha2.Egress) (net.IP, error) {
+	prevIP, prevIPPool, exists := c.getIPAllocation(egress.Name)
+	if exists {
+		// The Egress IP doesn't change, do nothing.
+		if prevIP.String() == egress.Spec.EgressIP && prevIPPool == egress.Spec.ExternalIPPool {
```

Comment:
When Egress IP is allocated from a pool, we do not write the IP back to the Spec right? Then we should check the case Spec.EgressIP == "" differently?

Reply:
We write the allocated IP back to the Egress's Spec, otherwise user and agent wouldn't know what IP it uses.

Reply:
But then we cannot differentiate the IP is from user or allocated by Antrea? Is it common in K8s to reuse a Spec attribute for generated state? I feel always good to differentiate desired state and generated state. Should we add the allocated IP to Status?

Reply:
> But then we cannot differentiate the IP is from user or allocated by Antrea? 

Do we need to differentiate them? both of them need to be marked as used in controller. agent doesn't care how the IP comes from.

> Is it common in K8s to reuse a Spec attribute for generated state?

The allocated ClusterIP and NodePort is in Service's Spec. User can request a specific ClusterIP when creating a service or let K8s allocate one. I think it's same for Egress's case. Similarly, user can specify nodeName of a Pod or let scheduler assign one. Besides, since user can request a specific EgressIP when specifying a Pool, it may make all code that consumes the Egress object have to iterate two fields to get the effective Egress IP if having allocated one in Status.

Reply:
Just from troubleshooting perspective, how users know the IP is from hirself or is allocated by Antrea? I know many systems clearly differentiate the desired state from users and the realized state by the system. But anyway, if this is a widely used pattern in K8s, I am fine with the approach too.

Reply:
Maybe we could also think it in this way: the Egress IP is still a desired state regardless of whether it's requested by user or allocated by system:
- I want these Pods to use this Egress IP when they access outside when I specify the IP. 
- I want these Pods to use any IP from the Pool when they access outside when I only specify the pool.

The IP allocation is more autofilled than a realized state, as having it in spec or status doesn't mean it has been realized in datapath. We could say the realized state comes from agents and is about whether the desired IP is realized on datapath or not.

Reply:
I think the allocated IPs should not be called desired state/intention. But anyway I would not be against to the approach as the same patterns exist in K8s.